### ISSUE TITLE
Reduce training epochs + check convergence speed sensitivity for Regression Tests.

### ISSUE THREAD

**Issue by kedz** [RasaHQ/rasa#8678]
**Description of Problem**: Currently our regression configs specify a longer training time than is strictly necessary. This is in part because they are being treated like a leaderboard which is a different thing.

Additionally, a bug recently escaped regression testing because the regression tests ran for too long. A commit unintentionally reduced the convergence speed so training to convergence took considerably longer. Since the max training epochs was too large though both the new commit and prior commit were eventually getting to the same test set result and so no issue was detected.

Ideally training would run for long enough to escape the random first few steps of training where parameters change alot but not necessarily need to be trained to convergence. This actual number needs to be determined empirically on the regression test datasets for both nlu and core.

**Overview of the Solution**: Determine appropriate training epochs for nlu and core configs/datasets, update configs in the training-data repo, and add a test to check for convergence speed

**Definition of Done**:

*   [ ] propose and align on convergence speed test with the rest of the research team
*   [ ] Create further issues for implementing convergence speed test as part of the regression tests, training-data configs are updated, [Notion documentation](https://www.notion.so/rasa/Implementing-Research-Features-897b2168b2c04652a38f135e4d8a2bad#79eb5efa40c9451a8e2885c7486e7abb) is updated to explain how to interpret the regression results, automating when these tests are run depending on what files have changed, etc.

**Comment by dakshvar22** [comment]
> Additionally, a bug recently escaped regression testing because the regression tests ran for too long. A commit unintentionally reduced the convergence speed so training to convergence took considerably longer. Since the max training epochs was too large though both the new commit and prior commit were eventually getting to the same test set result and so no issue was detected.

@kedz I don't think this explains the bug correctly. The models never converged during training because of the bug (even with large number of epochs). They just reported the same test accuracy numbers as before and hence weren't caught in the first place. The latter does not suggest that the model reached convergence.

My main question here is: Are we trying to come up with optimal number of epochs of every unique pair of (configuration, dataset)? 
If yes, that would be a lot of them. Do we plan to change how the test configuration comment is written to accommodate different number of epochs for each unique pair?
If not, there would still be some pairs for which training would run for more number of epochs than its optimal.

Thinking about the problem and the specific bug: The bug would have been caught if training accuracy was also reported as part of the results of model regression tests (for both on demand and scheduled regression tests). Making provision for that in the current workflow of model regression tests will be a win and will avoid the complexity of finding the optimal number of epochs for each unique pair.

What do you think?

**Comment by JEM-Mosig** [comment]
My take on this:
* Reducing the number of epochs only makes sense if a particular config converges on _all_ datasets at N epochs, but it is set to train at N + 100+. So I'd only look at the slowest configs (which ones are these?) and see if this is the case
* Daksh's comment about testing on the training set makes sense, I'd add this
* I think it would make sense to measure the convergence speed using tensorboard output somehow (e.g. estimate the slope of the accuracy/epoch graph, or noting at which epoch training accuracy was "sufficiently close" to final training accuracy)

**Comment by JEM-Mosig** [comment]
I've created issues to deal with adding training scores and convergence point estimates, which the infrastructure squad is going to deal with. Once these are done, adjusting the training epochs for the various models should be trivial, so I declare this issue as blocked until then.
